### ISSUE TITLE
Ask feature: plan publication

### ISSUE THREAD

**Issue by Sp3r4z** [Chocobozzz/PeerTube#455]
Hi
I've talk with someone who post videos on Youtube and Peertube, and this person said me that it's not possible, actually, to plan publication or videos so: here I am!

Is a publication planning feature is planned for the future or something like that?

Thanks for all you work :+1:

**Comment by Chocobozzz** [comment]
Just to be sure, when you talk about plan publication you mean changing the video privacy setting from `private` to `public` at a certain date and time?

**Comment by Sp3r4z** [comment]
Yeah, I think it can be a "simple" solution like that. Youtube seem to do the same as that. (Youtube have an "inetrval" value like "user with link only". I'm not an administrator of a Peertube instance, so I can't give more informations on that :s ).

But definitely a "changing video status `private`→`public` is a good idea :) (with of course a day and hours/minutes )

Thanks for your help.

**Comment by LecygneNoir** [comment]
Is it possible at the moment to do something like that with the API?
I am trying to script something that allow to edit a video from private to public at a given time, but I do not see method to update video inside API.

Have I missed something?

Thanks!

**Comment by Sp3r4z** [comment]
I've checked it at the doc/API, and it seems that no API is provided for that purpose. We can check and GET the privacies for videos, but not for a specific one… so it's seem we are trapped, and it's not actually possible. I've no idea if it's something simple, adding this feature into the API, but it's definitely a good idea for future.
Maybe like Youtube, with three states: public, unlisted (people with link) and private.

Thanks a lot for all you work.

**Comment by Sp3r4z** [comment]
@Chocobozzz Doc give "Public Unlisted" not Private status. (maybe just a missing point)

But, with that, maybe it's possible to implement something like: 
Public on [YYYY/MM/DD][HH:MM] during upload (with a date/time picker)

Maybe it's also about a CRON task, for instance admins.

**Comment by Chocobozzz** [comment]
No, it's not a missing point: https://github.com/Chocobozzz/PeerTube/issues/317

**Comment by Sp3r4z** [comment]
@Chocobozzz seems really logical then. Thanks for your answer.

I've seen on wiki pages, script for uploading videos on Youtube and Peertube simultaneously, so I think it can be a really cool idea (scheduling publication).

We'll see then :)

**Comment by Sp3r4z** [comment]
@LecygneNoir I think CRON is the only solution for that, a CRON task on the server which verify every minute (10, 15, 30mins… don't know) if one or more video have to be publish, and just change (as doc say) publishing status from private/unlisted to public. And with that, it can be automatic.

Let's see :)

**Comment by LecygneNoir** [comment]
Yep it is a reasonable options if you have access to the Peertube server, and thus to solve this issue, I even think the future PublishAt fields for videos could help, for example by allowing to set a publishat date in the future, and change status according to this date :-)

However for the moment I am very too bad in typescript to do something like that sorry (mainly, my tries just broke the entire code ^^')
For an external application like Prismedia, consider you have no access to the Peertube server, so cron inside the server is not an option, I am working on that.

**Comment by Sp3r4z** [comment]
@LecygneNoir I'm not on the "Prismedia" repository, isn't it? So I ask for exactly what you talk about, in your first part of the message.
Imho this should be doing by something else than the server… not a third-part script of else!

Hope it'll be done soon, because I've heard a lot about that difference with "older video hosting services", which one is really stalling for some user :disappointed:

**Comment by LecygneNoir** [comment]
There are at least 2 more difficulties on this issue, the timezone problem and the "frequency" problem.

1) When we plan a video, it's hard to guess the user wants to use the local server timezone (which he does not know, and which could be misconfigured) or his local timezone. 
I do not know if it is possible to check the user timezone perhaps through the browser timezone, something like that 🤔 

Youtube offers to choose timezone when you plan the video, I guess something like that could be possible, but we will need to compute it after that so the server know when to publish the video according to its own timezone...

2) About the "frequency" problem, if we use cron, we need to tell when we run the check for "video to publish", either each minute, each 5 minutes, etc. On large instances, checking each minutes if videos are ready to be published could cost lot of resources? 

Relying on the atd daemon could help, as it is possible to tell it "at this time, do something", but it adds a dependencies, and it is not very easy to use.

Addition to that, we need to deal with the fact someone could plan a video in the past, or that the date could be passed when the video ends its processing, if the processing take long time. 

Not so easy to solve, but open to discussion!

**Comment by Sp3r4z** [comment]
For your first point, it's really interesting and complex… we have to find a way to "translate" from a human readable time to a timestamp… not so easy :disappointed: 

About your second point:
- Video can't be publish before the processing, maybe the simplest solution is that: when the video is processed, activate the possibility to plan it. Maybe doing a computation with the video size + Xhours (or 1day), something like that.
- I've thought about something like 15minutes or 30minutes. It can be a good think, no-one want to plan a video at "2:47" (I mean, it's really funny but… hum…). So the CRON and scheduling system can be like that.

It's complicated, but I think we can find a suitable solution :)

**Comment by DavidLibeau** [comment]
Hi, this feature is wanted by videomakers. Do we have any plan about integrating it in PeerTube? Thanks
